server.py

```diff
@@ -25,39 +25,44 @@ def add_datum(self, datum):
                                                          time.time())
 
     def get_datum(self):
-        datum = {}
-        for i in range(len(self._calculators)):
-            datum.update({f"patient-{i}": self._calculators[i].descriptors})
-        return datum
+        return {f"patient-{i}": self._calculators[i].descriptors
+                for i in range(len(self._calculators))}
 
 
 class Communicator():
 
     def __init__(self, port=5000):
-        self._socket = zmq.Context().socket(zmq.PUB)
+        self._context = zmq.Context()
+        self._socket = self._context.socket(zmq.PUB)
         self._socket.bind(f"tcp://*:{port}")
 
+    def close(self):
```

Comment:
What happens if `close` is called twice?

Reply:
Ping.

Reply:
I commented on this before, but it seems like it didn't show up for you.  Nothing happens if you close twice.  It's just closed.

Reply:
Fair enough. (And polite of the zmq library to provide that behavior.) A common pattern in situations like this is to make sure you only close a resource once, just to guard against aberrant behaviors from unpolite libraries.